service/history/processing/interface.go

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination interface_mock.go -self_package github.com/uber/cadence/service/history/processing
+
+package processing
+
+import (
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/service/history/task"
+)
+
+type (
```

Comment:
I think the "Job" name is a bit too generic and doesn't imply its actual responsibilities. What would you think about renaming it to something like VirtualQueue, TaskQueue or [Scoped|Filtered]Queue? I like VirtualQueue the most.

If we use "VirtualQueue" name instead of "Job", I'd suggest renaming the interfaces as follows: 

- JobInfo: VirtualQueueScope
- Job: VirtualQueue
- JobSplitPolicy: VirtualQueueSplitPolicy
- JobQueue: VirtualQueueCollection maybe? (I'm less confident about this one..)
- JobQueueManager: VirtualQueueManager

Reply:
I still want to keep `Info` in the type name as I think it contains more than just a scope. If the metadata only has min/max level and domain filter, then I think `Scope` is a perfect name. But in our case, it also contains information like read level, ack level, ID, etc. So it more like a status, instead of just scope. So maybe `xxInfo` or `xxStatus`?

In terms of the name of `Job`, I am not sure either. I am a little bit hesitate to rename it to `Queue` because then we don't have a good name for the current `JobQueue`. From my perspective, Job is one level below the concept of queue, more like a sub-task. 

If possible, can you elaborate a little bit on why you think `Job` doesn't imply its responsibilities? For example, which part is missing. Maybe based on that we can came up with a better name.

Reply:
Synced offline. Will use the following names:
- ProcessingQueue
- ProcessingQueueState
- ProcessingQueueSplitPolicy
- ProcessingQueueCollection
- ProcessingQueueManager